docs/plugins/npe2_manifest_specification.md

```diff
@@ -0,0 +1,578 @@
+    ...
+```
+
+**Parameters**
+
+path(str)
+: Path to file, directory, or resource (like a url).
+
+layer_data(list of LayerData)
+: Each `LayerData` element is a tuple of `(data,meta,layer_type)`.
+
+**Return**
+
+Returns a list of paths that were successfully written.
+
+#### single-layer writer
+
+```python
+def writer_function(
+    path: str, data, meta
+) -> Optional[str]:
+    ...
+```
+
+**Parameters**
+
+path(str)
+: Path to file, directory, or resource (like a url).
+
+data(array or list of array) : Layer data. Can be N dimensional. If an image
+and meta[‘rgb’] is True then the data should be interpreted as RGB or RGBA. If
+meta[‘multiscale’] is True, then the data should be interpreted as a
+multiscale image.
+
+meta(dict)
+: Layer metadata.
+
+**Return**
+
+If data is successfully written, return the path that was written. Otherwise,
+if nothing was done, return None.
+
+## Sample Data
+
+There are two ways of specifying sample data.
+
+1. As a **sample data generator**, a function that returns layer data.
+2. As a **sample data uri**, a `uri` that should be read using a `reader`
+   plugin.
+
+### Required fields
+
+_sample data generator_
+
+- **command** Identifier of a command that returns a layer data tuple.
+- **display_name** String to show in the GUI when referring to this sample.
+- **key** A key to identify this sample. Must be unique across the samples
+  provided by a single plugin.
+
+_sample data uri_
+
+- **uri** Path or URL to a data resource.
+- **display_name** String to show in the GUI when referring to this sample.
+- **key** A key to identify this sample. Must be unique across the samples
+  provided by a single plugin.
+
+### Optional fields
+
+_sample data uri_
+
+- **reader_plugin** Name of plugin to use to open the `uri`.
+
+### Examples
+
+**Example**
+
+Sample data generator
+
+```yaml
+name: napari
+contributions:
+  commands:
+    - id: napari.data.astronaut
+      title: Generate astronaut sample
+      python_name: napari.plugins._skimage_data:astronaut
+  sample_data:
+    - display_name: Astronaut (RGB)
+      key: astronaut
+      command: napari.data.astronaut
+```
+
+**Example**
+
+Sample data uri
+
+```yaml
+name: my-sample
+contributions:
+  sample_data:
+    - display_name: Tabueran Kribati
+      key: napari
+      uri: https://en.wikipedia.org/wiki/Napari#/media/File:Tabuaeran_Kiribati.jpg
+```
+
+### Calling convention
+
+_sample data generator_
+
+```python
+def sample_data_generator()->List[LayerData]:
+  ...
+```
+
+## Widgets
+
+```{note}
+The `napari_experimental_provide_dock_widget` and
+`napari_experimental_provide_function` hook implementations may be bound as the
+callable for Widget contributions.  `napari_experimental_provide_function`
+should use `autogenerate: true`.  See more details below.
+```
+
+### Required fields
+
+- **command** Identifier of a command that returns a Widget instance.
+- **display_name** User-facing name to use for the widget in, for example,
```

Comment:
it's still weird to me when I get down here that `python_name` is not required for these but they do nothing unless it's provided... Is there some way we could disambiguate this somehow? Maybe the comment I made above is enough but idk...

Reply:
Yes, my comment above about python-name and register command dynamically being alternatives should be mentioned here.

If someone omits python name, and doesn't register anything in their activate function, it will raise an error at runtime when the command is finally called

Reply:
added a note here and to the command docs.